clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -5024,11 +5023,24 @@ bool SYCLIntegrationFooter::emit(raw_ostream &OS) {
   Policy.SuppressTypedefs = true;
   Policy.SuppressUnwrittenScope = true;
 
+  llvm::SmallSet<const VarDecl *, 8> VisitedSpecConstants;
+
   // Used to uniquely name the 'shim's as we generate the names in each
   // anonymous namespace.
   unsigned ShimCounter = 0;
   for (const VarDecl *VD : SpecConstants) {
     VD = VD->getCanonicalDecl();
+
+    // Skip if this isn't a SpecIdType.  This can happen if it was a deduced
+    // type.
+    if (!Util::isSyclSpecIdType(VD->getType().getCanonicalType()))
```

Comment:
Do we always want `isSyclSpecIdType()` to inspect the canonical type (e.g., sink the calls to `getCanonicalType()` into the definition of the function rather than letting the caller decide which to pass)?

Reply:
I think we do, yes.  All we care about is whether this is a spec-id type, we don't really care about whether it is wrapped in typedefs or not.

Reply:
Okay, then I'd change the implementation of `isSyclSpecIdType()` to do the canonicalization rather than adding the calls here. I don't care if that's a follow-up or done as part of this patch though. WDYT?

Reply:
You're right here, we probably want the entire list of isNNNType functions in Util to canonicalize.

Reply:
https://github.com/intel/llvm/pull/3916